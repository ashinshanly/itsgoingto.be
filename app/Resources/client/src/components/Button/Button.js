import React from 'react'
import PropTypes from 'prop-types'
import EventBus from 'components/EventBus'
import Spinner from 'components/Spinner/Spinner'
import './Button.scss'

class Button extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      disabled : false,
      loading  : false
    }
  }

  componentDidMount = () => {
    this._mounted = true
    this.eventBus = EventBus.getEventBus()
    if (this.props.submitEvent) {
      this.eventListener = this.eventBus.addListener(this.props.submitEvent, () => {
        this.handlePress()
      })
    }
  }

  componentWillUnmount = () => {
    this._mounted = false
    if (this.eventListener) {
      this.eventListener.remove()
    }
  }

  isDisabled = () => this.props.disabled || this.state.disabled

  handlePress = (event) => {
    if (event) {
      event.preventDefault()
    }
    if (!this.isDisabled()) {
      this.setState({
        disabled : true,
        loading  : true
      })
      this.props.callback().then((reset) => {
        if (reset !== false && this._mounted) {
          this.setState({
            disabled : false,
            loading  : false
          })
        }
      })
    }
  }

  render = () => (
    <button
      className={'btn ' + this.props.className + (this.isDisabled() ? ' disabled' : '')}
      disabled={this.props.disabled}
      onClick={this.handlePress}>
      {this.state.loading && <Spinner />}
      {!this.state.loading && <span>{this.props.text}</span>}
    </button>
  )
}

Button.propTypes = {
  text        : PropTypes.string.isRequired,
  className   : PropTypes.string,
  disabled    : PropTypes.bool,
  callback    : PropTypes.func.isRequired,
  submitEvent : PropTypes.string
}

Button.defaultProps = {
  disabled : false,
}

export default Button
