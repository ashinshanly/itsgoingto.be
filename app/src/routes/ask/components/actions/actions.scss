@import "variables";

.actions {
  margin-top: 2em;

  @media(max-width: $screen-md-max) {
    font-size: 1.3em;
  }

  @media(max-width: $screen-xs-max) {
    text-align: center;
    margin-top: 4em;
  }

  .pull-right {
    @media(max-width: $screen-xs-max) {
      float: none;
    }
  }
}
