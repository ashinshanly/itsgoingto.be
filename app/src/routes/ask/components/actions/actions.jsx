import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { browserHistory } from 'react-router'
import { postPoll } from 'services/api'
import Button from 'components/button'
import './actions.scss'

export class Actions extends React.Component {
  submit = () => this.props.postPoll()
    .then((response) => {
      if (response !== false) {
        browserHistory.push('/' + response.identifier)
        return false
      }
      return true
    })

  render () {
    const { hasQuestion, canSubmitPoll } = this.props

    return (
      <div>
        <div className={'actions hideable' + (hasQuestion ? '' : ' gone')}>
          <Button className='pull-right' text='Create Poll' disabled={!canSubmitPoll} callback={this.submit} />
        </div>
      </div>
    )
  }
}

Actions.propTypes = {
  hasQuestion   : PropTypes.bool.isRequired,
  canSubmitPoll : PropTypes.bool.isRequired,
  postPoll      : PropTypes.func.isRequired,
}

const mapDispatchToProps = (dispatch) => ({
  postPoll : () => dispatch(postPoll())
})

export default connect(null, mapDispatchToProps)(Actions)
