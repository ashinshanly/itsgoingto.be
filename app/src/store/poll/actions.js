import { prop, compose, omit, map, ifElse, isNil, merge, __ } from 'ramda'
import { hasQuestionSelector } from 'store/poll/selectors'
import { addAnswer, clearAnswers, updateAnswers } from 'store/answers/actions'

// ------------------------------------
// Constants
// ------------------------------------
export const POLL_UPDATE     = 'POLL_UPDATE'
export const POLLS_SET       = 'POLLS_SET'
export const POLL_PAGE_SET   = 'POLL_PAGE_SET'
export const POLL_COUNT_SET  = 'POLL_COUNT_SET'
export const QUESTION_UPDATE = 'QUESTION_UPDATE'

// ------------------------------------
// Actions
// ------------------------------------
/**
 * Update the poll in the state then dispatch to update the answers
 *
 * @param  {Poll}     poll The poll to add/update the state with
 *
 * @return {Function}      redux-thunk callable function
 */
export const updatePoll = (poll) => (dispatch) =>
  ifElse(
    compose(isNil, (prop('answers'))),
    (poll) => Promise.resolve(
      dispatch({
        type : POLL_UPDATE,
        poll : omit(['answers'])(poll)
      })
    ).then(() => poll),
    (poll) => Promise.all([
      dispatch({
        type : POLL_UPDATE,
        poll : omit(['answers'])(poll)
      }),
      dispatch(updateAnswers(prop('answers', poll)))
    ]).then(() => poll)
  )(poll)

/**
 * Set the polls in the state
 *
 * @param  {Poll[]}   polls The array of polls to put in the state
 *
 * @return {Function}       dispatchable object
 */
export const setPolls = (polls = []) => ({
  type  : POLLS_SET,
  polls : map(omit(['answers']))(polls),
})

/**
 * Set the poll page in the state
 *
 * @param  {integer}  count The total count of polls to put in the state
 *
 * @return {Function}       dispatchable object
 */
export const setPollPage = (page) => ({
  type  : POLL_PAGE_SET,
  page,
})

/**
 * Set the poll count in the state
 *
 * @param  {integer}  count The total count of polls to put in the state
 *
 * @return {Function}       dispatchable object
 */
export const setPollCount = (count) => ({
  type  : POLL_COUNT_SET,
  count,
})

/**
 * Update the question text in the state for a given poll
 * Dispatch to insert or clear the answers appropriately
 *
 * @param  {string}   text       The question text
 * @param  {string}   identifier The poll identifier to update
 *
 * @return {Function}             redux-thunk callable function
 */
export const updateQuestion = (text = '', identifier = '') => (dispatch, getState) => {
  const hadQuestion = hasQuestionSelector(getState(), identifier)

  dispatch({
    type       : QUESTION_UPDATE,
    question   : text,
    identifier
  })
  const hasQuestion = hasQuestionSelector(getState(), identifier)

  if (hasQuestion && !hadQuestion) {
    dispatch(addAnswer())
  } else if (hadQuestion && !hasQuestion) {
    dispatch(clearAnswers())
  }

  return Promise.resolve()
}

/**
 * Update the responses for the poll
 *
 * @param  {object}   responses  The responses object
 * @param  {string}   identifier The poll identifier to update
 *
 * @return {Function}            redux-thunk callable function
 */
export const updateResponses = (responses, identifier) => (dispatch) => Promise.all([
  dispatch({
    type : POLL_UPDATE,
    poll : compose(omit(['answers']), merge(__, { identifier }))(responses)
  }),
  dispatch(updateAnswers(prop('answers', responses)))
]).then(() => responses)
