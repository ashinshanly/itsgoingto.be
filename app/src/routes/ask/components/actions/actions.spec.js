/* global expect, jest */
import React from 'react'
import { browserHistory } from 'react-router'
import { shallow } from 'enzyme'
import Button from 'components/button'
import { Actions } from './actions'

const props = {
  hasQuestion: true,
  canSubmitPoll: true,
  postPoll: jest.fn(() => Promise.resolve({ identifier: 'jdH93HS' }))
}
let wrapper

describe('(Route) Ask', () => {
  describe('(Component) Actions', () => {
    describe('(Action) Create Poll click', () => {
      beforeEach(() => {
        wrapper = shallow(<Actions {...props} />)
        browserHistory.push = jest.fn()
      })

      it('should call postPoll', () => {
        wrapper.find(Button).props().callback()
        expect(props.postPoll).toBeCalled()
      })

      it('should redirect with poll identifier on successful postPoll and return false', () => {
        return wrapper.find(Button).props().callback()
          .then((response) => {
            expect(response).toBe(false)
            expect(browserHistory.push).toBeCalledWith('/jdH93HS')
          })
      })

      it('should not redirect failed postPoll and return true', () => {
        wrapper = shallow(<Actions {...props} postPoll={jest.fn(() => Promise.resolve(false))} />)

        return wrapper.find(Button).props().callback()
          .then((response) => {
            expect(response).toBe(true)
            expect(browserHistory.push).not.toBeCalled()
          })
      })
    })

    describe('(Render)', () => {
      describe('when hasQuestion is true', () => {
        wrapper = shallow(<Actions {...props} hasQuestion={true} />)
        it('matches snapshot', () => {
          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('when hasQuestion is false', () => {
        wrapper = shallow(<Actions {...props} hasQuestion={false} />)
        it('matches snapshot', () => {
          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('when canSubmitPoll is true', () => {
        wrapper = shallow(<Actions {...props} canSubmitPoll={true} />)
        it('matches snapshot', () => {
          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('when canSubmitPoll is false', () => {
        wrapper = shallow(<Actions {...props} canSubmitPoll={false} />)
        it('matches snapshot', () => {
          expect(wrapper).toMatchSnapshot()
        })
      })
    })
  })
})