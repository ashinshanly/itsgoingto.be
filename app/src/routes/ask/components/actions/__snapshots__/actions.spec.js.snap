// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Route) Ask (Component) Actions (Render) when canSubmitPoll is false matches snapshot 1`] = `
<div>
  <div
    className="actions hideable"
  >
    <Button
      callback={[Function]}
      className="pull-right"
      disabled={false}
      submitEvent={null}
      text="Create Poll"
    />
  </div>
</div>
`;

exports[`(Route) Ask (Component) Actions (Render) when canSubmitPoll is true matches snapshot 1`] = `
<div>
  <div
    className="actions hideable"
  >
    <Button
      callback={[Function]}
      className="pull-right"
      disabled={false}
      submitEvent={null}
      text="Create Poll"
    />
  </div>
</div>
`;

exports[`(Route) Ask (Component) Actions (Render) when hasQuestion is false matches snapshot 1`] = `
<div>
  <div
    className="actions hideable"
  >
    <Button
      callback={[Function]}
      className="pull-right"
      disabled={false}
      submitEvent={null}
      text="Create Poll"
    />
  </div>
</div>
`;

exports[`(Route) Ask (Component) Actions (Render) when hasQuestion is true matches snapshot 1`] = `
<div>
  <div
    className="actions hideable"
  >
    <Button
      callback={[Function]}
      className="pull-right"
      disabled={false}
      submitEvent={null}
      text="Create Poll"
    />
  </div>
</div>
`;
